### ISSUE TITLE
Quality of reconstruction drops after epoch 4 and advice on specific parameters to adjust

### ISSUE THREAD

**Issue by wsalmen2** [ml-struct-bio/drgnai#12]
Hi Michal, 

I have been able to successfully run a job and noticed that after epoch 4 my reconstruction goes from a reasonable quality map to a slight blob on epoch 5 and then it looks terrible all the way to epoch 100. I am currently working with a little over 100,000 particles downsampled to ~3.5 A/pix. Do you have any advice on which parameters are worth adjusting to improve the training quality? 

Also, is there a reason z_dim is default to 4? I know in cryoDrgn it's default is 8. Does this play a role in HPS during the first 5 epochs? Any advice is greatly appreciated. 

Best,
Wil

**Comment by michal-g** [comment]
Hey Wil, can you share the command and parameters you used when getting this result to make it easier to figure out what might be going on?

There is no particular reason for the default of `z_dim: 4` for drgn-AI and `z_dim: 8` for cryoDRGN, although we will likely harmonize both to `z_dim: 8` at some point in the future. The value of 8 is better for mixed precision speedups implemented  using [torch.cuda.amp](https://pytorch.org/docs/stable/amp.html), which has been included in the cryoDRGN software but not (yet) applied to drgn-AI, while 4 may in some cases prevent underfitting of points in the latent space (see e.g. the [curse of dimensionality](https://en.wikipedia.org/wiki/Curse_of_dimensionality)).

That being said, we encourage users to try both values whenever possible to better understand what might be going on with their data. `z_dim` is playing some role in your results during the HPS training phase, unless you are using something other than the default value for `pose_only_phase: 0`, but it is unlikely to be decisive.

Anyway, as a first guess I would suggest increasing the model size, as the default parameters may not be sufficient to capture the complexity present in your dataset, especially when moving into SGD mode which would be happening at around the 5th epoch if you are using the default value of `n_imgs_pose_search: 500000`. The default is `hypervolume_dim: 256` so perhaps try `hypervolume_dim: 512` or `hypervolume_dim: 1024` as your compute allows!

-Michal

**Comment by wsalmen2** [comment]
Hi Michal,

Here is my config.ymal:

particles: /path_to_particles/particles_256.mrcs
ctf: /path_to_ctf/ctf.pkl
pose: null
quick_config:
  capture_setup: spa
  reconstruction_type: het
  pose_estimation: abinitio
  conf_estimation: autodecoder
lazy: true
batch_size_sgd: 32

I am currently rerunning my job with hypervolume_dim: 1024 and z_dim: 8, since these were the values I have previously used in cryoDrgn. 

Alternatively, would it make sense to try rerunning the job with pose_estimation: refine and provide the pose.4.pkl from the epoch 4?

Thanks,
Wil

**Comment by mpm896** [comment]
Hi Michal,

I'm chiming in because I'm facing the exact same issue. I a get decent hypervolume reconstruction (if I'm correctly understanding what the hypervolume is, the output `reconstruction.-1.mrc` and `reconstruction.0.mrc`) which turns to an unrecognizable blob after 1-2 SGD epochs. Using similar params as @wsalmen2 except a `batch_size_sgd` of 128. @wsalmen2  did you get better results from the SGD phase of training after changing `hypervolume_dim: 1024` and `z_dim: 8`? 

I also noticed that the learning curve shows a large spike once it switches to SGD (attached below). Has this been the case for your successful training runs? I was wondering if this could indicate too high of a learning rate once switching to SGD. I'm not sure which learning rate to mess with in this case. Thanks!

<img width="649" alt="Image" src="https://github.com/user-attachments/assets/ca4bd952-a568-4b9f-b771-1ef09c65be60" />

**Comment by mpm896** [comment]
Just to update, this spike at the switch from HPS to SGD lead me to thinking that the learning rates were too high. I've reduced the `lr_pose_table` to `1e-4` and `lr_conf_table` to `1e-4` and now the hypervolumes that are output every 5 epochs are very good. I haven't tried increasing `z_dim` or `hypervolume_dim` yet, but will try that soon as well. It's also worth noting that I'm working on a <200 kDa complex.
